File: sdk/assemblyscript/src/assembly/vectors.ts
/*
 * Copyright 2024 Hypermode, Inc.
 * Licensed under the terms of the Apache License, Version 2.0
 * See the LICENSE file that accompanied this code for further details.
 *
 * SPDX-FileCopyrightText: 2024 Hypermode, Inc. <hello@hypermode.com>
 * SPDX-License-Identifier: Apache-2.0
 */

function assertEqualLength<T extends number>(a: T[], b: T[]): void {
  if (a.length !== b.length) {
    throw new Error("Vectors must be the same length.");
  }
}

function checkValidArray<T extends number>(a: T[]): void {
  if (a.length === 0) {
    throw new Error("Vector must not be empty.");
  }

  checkValidNumber(a[0]);
}

export function checkValidNumber<T extends number>(a: T): void {
  if (!isInteger(a) && !isFloat(a) && !isSigned(a)) {
    throw new Error("Vector must contain numbers.");
  }
}
/**
 *
 * Add two vectors together, returning a new vector.
 * @param a: The first vector
 * @param b: The second vector
 * @returns: The sum of the two vectors
 */
export function add<T extends number>(a: T[], b: T[]): T[] {
  assertEqualLength(a, b);
  checkValidArray(a);
  checkValidArray(b);
  const result = new Array<T>(a.length);
  for (let i = 0; i < a.length; i++) {
    result[i] = (a[i] + b[i]) as T;
  }
  return result;
}

/**
 *
 * Add two vectors together, modifying the first vector.
 * @param a: The first vector
 * @param b: The second vector
 */
export function addInPlace<T extends number>(a: T[], b: T[]): void {
  assertEqualLength(a, b);
  checkValidArray(a);
  checkValidArray(b);
  for (let i = 0; i < a.length; i++) {
    a[i] = (a[i] + b[i]) as T;
  }
}

/**
 *
 * Subtract two vectors, returning a new vector.
 * @param a: The first vector
 * @param b: The second vector
 * @returns: The difference of the two vectors
 */
export function subtract<T extends number>(a: T[], b: T[]): T[] {
  assertEqualLength(a, b);
  checkValidArray(a);
  checkValidArray(b);
  const result = new Array<T>(a.length);
  for (let i = 0; i < a.length; i++) {
    result[i] = (a[i] - b[i]) as T;
  }
  return result;
}

/**
 *
 * Subtract two vectors, modifying the first vector.
 * @param a: The first vector
 * @param b: The second vector
 */
export function subtractInPlace<T extends number>(a: T[], b: T[]): void {
  assertEqualLength(a, b);
  checkValidArray(a);
  checkValidArray(b);
  for (let i = 0; i < a.length; i++) {
    a[i] = (a[i] - b[i]) as T;
  }
}

/**
 *
 * add a number to a vector, returning a new vector.
 * @param a: The first vector
 * @param b: The number to add
 * @returns: the result vector, with the number added to each element
 */
export function addNumber<T extends number>(a: T[], b: T): T[] {
  checkValidArray(a);
  checkValidNumber(b);
  const result = new Array<T>(a.length);
  for (let i = 0; i < a.length; i++) {
    result[i] = (a[i] + b) as T;
  }
  return result;
}

/**
 *
 * add a number to a vector, modifying the first vector.
 * @param a: The first vector
 * @param b: The number to add
 */
export function addNumberInPlace<T extends number>(a: T[], b: T): void {
  checkValidArray(a);
  checkValidNumber(b);
  for (let i = 0; i < a.length; i++) {
    a[i] = (a[i] + b) as T;
  }
}

/**
 *
 * Subtract a number from a vector, returning a new vector.
 * @param a: The first vector
 * @param b: The number to subtract
 * @returns: the result vector, with the number subtracted from each element
 */
export function subtractNumber<T extends number>(a: T[], b: T): T[] {
  checkValidArray(a);
  checkValidNumber(b);
  const result = new Array<T>(a.length);
  for (let i = 0; i < a.length; i++) {
    result[i] = (a[i] - b) as T;
  }
  return result;
}

/**
 *
 * Subtract a number from a vector, modifying the first vector.
 * @param a: The first vector
 * @param b: The number to subtract
 */
export function subtractNumberInPlace<T extends number>(a: T[], b: T): void {
  checkValidArray(a);
  checkValidNumber(b);
  for (let i = 0; i < a.length; i++) {
    a[i] = (a[i] - b) as T;
  }
}

/**
 * Multiple numbers to a vector, returning a new vector.
 * @param a: The first vector
 * @param b: The number to multiply
 * @returns: the result vector, with the number multiplied to each element
 */
export function multiplyNumber<T extends number>(a: T[], b: T): T[] {
  checkValidArray(a);
  checkValidNumber(b);
  const result = new Array<T>(a.length);
  for (let i = 0; i < a.length; i++) {
    result[i] = (a[i] * b) as T;
  }
  return result;
}

/**
 *
 * Multiply a number to a vector, modifying the first vector.
 * @param a: The first vector
 * @param b: The number to multiply
 */
export function multiplyNumberInPlace<T extends number>(a: T[], b: T): void {
  checkValidArray(a);
  checkValidNumber(b);
  for (let i = 0; i < a.length; i++) {
    a[i] = (a[i] * b) as T;
  }
}

/**
 *
 * Divide a number from a vector, returning a new vector.
 * @param a: The first vector
 * @param b: The number to divide
 * @returns: the result vector, with the number divided from each element
 */
export function divideNumber<T extends number>(a: T[], b: T): T[] {
  checkValidArray(a);
  checkValidNumber(b);
  const result = new Array<T>(a.length);
  for (let i = 0; i < a.length; i++) {
    result[i] = (a[i] / b) as T;
  }
  return result;
}

/**
 *
 * Divide a number from a vector, modifying the first vector.
 * @param a: The first vector
 * @param b: The number to divide
 */
export function divideNumberInPlace<T extends number>(a: T[], b: T): void {
  checkValidArray(a);
  checkValidNumber(b);
  for (let i = 0; i < a.length; i++) {
    a[i] = (a[i] / b) as T;
  }
}

/**
 * Calculate the dot product of two vectors.
 * @param a: The first vector
 * @param b: The second vector
 * @returns: The dot product of the two vectors
 */
export function dot<T extends number>(a: T[], b: T[]): T {
  checkValidArray(a);
  checkValidArray(b);
  assertEqualLength(a, b);
  let result: number = 0;
  for (let i = 0; i < a.length; i++) {
    result += a[i] * b[i];
  }
  return result as T;
}

/**
 * Calculate the magnitude of a vector.
 * @param a: The vector
 * @returns: The magnitude of the vector
 */
export function magnitude<T extends number>(a: T[]): f64 {
  checkValidArray(a);
  return sqrt<f64>(dot(a, a));
}

/**
 * Calculate the cross product of two 3D vectors.
 * @param a: The first vector
 * @param b: The second vector
 * @returns: The cross product of the two vectors
 */
export function normalize<T extends number>(a: T[]): f64[] {
  checkValidArray(a);
  const magnitudeValue = magnitude(a);
  const result: f64[] = new Array<f64>(a.length);
  for (let i = 0; i < a.length; i++) {
    result[i] = (a[i] as f64) / magnitudeValue;
  }
  return result;
}

/**
 *
 * Calculate the sum of a vector.
 * @param a: The vector
 * @returns: The sum of the vector
 */
export function sum<T extends number>(a: T[]): T {
  checkValidArray(a);
  let result: number = 0;
  for (let i = 0; i < a.length; i++) {
    result += a[i];
  }
  return result as T;
}

/**
 *
 * Calculate the product of a vector.
 * @param a: The vector
 * @returns: The product of the vector
 */
export function product<T extends number>(a: T[]): T {
  checkValidArray(a);
  let result: number = 1;
  for (let i = 0; i < a.length; i++) {
    result *= a[i];
  }
  return result as T;
}

/**
 *
 * Calculate the mean of a vector.
 * @param a: The vector
 * @returns: The mean of the vector
 */
export function mean<T extends number>(a: T[]): f64 {
  checkValidArray(a);
  return f64(sum(a)) / f64(a.length);
}

/**
 *
 * Calculate the median of a vector.
 * @param a: The vector
 * @returns: The median of the vector
 */
export function min<T extends number>(a: T[]): T {
  checkValidArray(a);
  let result = a[0];
  for (let i = 1; i < a.length; i++) {
    if (a[i] < result) {
      result = a[i];
    }
  }
  return result;
}

/**
 *
 * Calculate the maximum of a vector.
 * @param a: The vector
 * @returns: The maximum of the vector
 */
export function max<T extends number>(a: T[]): T {
  checkValidArray(a);
  let result = a[0];
  for (let i = 1; i < a.length; i++) {
    if (a[i] > result) {
      result = a[i];
    }
  }
  return result;
}

/**
 *
 * Calculate the absolute value of a vector.
 * @param a: The vector
 * @returns: The absolute value of the vector
 */
export function abs<T extends number>(a: T[]): T[] {
  checkValidArray(a);
  const result = new Array<T>(a.length);
  for (let i = 0; i < a.length; i++) {
    result[i] = a[i] < 0 ? (-a[i] as T) : a[i];
  }
  return result;
}

/**
 *
 * Calculate the absolute value of a vector, modifying the first vector.
 * @param a: The vector
 */
export function absInPlace<T extends number>(a: T[]): void {
  checkValidArray(a);
  for (let i = 0; i < a.length; i++) {
    a[i] = a[i] < 0 ? (-a[i] as T) : a[i];
  }
}

/**
 *
 * Calculate the euclidian distance between two vectors.
 * @param a: The first vector
 * @param b: The second vector
 * @returns: The euclidian distance between the two vectors
 */
export function euclidianDistance<T extends number>(a: T[], b: T[]): f64 {
  checkValidArray(a);
  let sum: number = 0;
  for (let i = 0; i < a.length; i++) {
    sum += (a[i] - b[i]) ** 2;
  }
  return sqrt<f64>(sum);
}

File: sdk/go/pkg/vectors/vectors.go
/*
 * Copyright 2024 Hypermode Inc.
 * Licensed under the terms of the Apache License, Version 2.0
 * See the LICENSE file that accompanied this code for further details.
 *
 * SPDX-FileCopyrightText: 2024 Hypermode Inc. <hello@hypermode.com>
 * SPDX-License-Identifier: Apache-2.0
 */

package vectors

import (
	"math"

	"golang.org/x/exp/constraints"
)

// Add adds two vectors together and returns the result.
func Add[T constraints.Integer | constraints.Float](a, b []T) []T {
	assertEqualLength(a, b)
	result := make([]T, len(a))
	for i := range a {
		result[i] = a[i] + b[i]
	}
	return result
}

// AddInPlace adds two vectors together and stores the result in the first vector.
func AddInPlace[T constraints.Integer | constraints.Float](a, b []T) {
	assertEqualLength(a, b)
	for i := range a {
		a[i] += b[i]
	}
}

// Subtract subtracts one vector from another and returns the result.
func Subtract[T constraints.Integer | constraints.Float](a, b []T) []T {
	assertEqualLength(a, b)
	result := make([]T, len(a))
	for i := range a {
		result[i] = a[i] - b[i]
	}
	return result
}

// SubtractInPlace subtracts one vector from another and stores the result in the first vector.
func SubtractInPlace[T constraints.Integer | constraints.Float](a, b []T) {
	assertEqualLength(a, b)
	for i := range a {
		a[i] -= b[i]
	}
}

// AddNumber adds a number to each element of a vector and returns the result.
func AddNumber[T constraints.Integer | constraints.Float](a []T, b T) []T {
	result := make([]T, len(a))
	for i := range a {
		result[i] = a[i] + b
	}
	return result
}

// AddNumberInPlace adds a number to each element of a vector and stores the result in the vector.
func AddNumberInPlace[T constraints.Integer | constraints.Float](a []T, b T) {
	for i := range a {
		a[i] += b
	}
}

// SubtractNumber subtracts a number from each element of a vector and returns the result.
func SubtractNumber[T constraints.Integer | constraints.Float](a []T, b T) []T {
	result := make([]T, len(a))
	for i := range a {
		result[i] = a[i] - b
	}
	return result
}

// SubtractNumberInPlace subtracts a number from each element of a vector and stores the result in the vector.
func SubtractNumberInPlace[T constraints.Integer | constraints.Float](a []T, b T) {
	for i := range a {
		a[i] -= b
	}
}

// MultiplyNumber multiplies each element of a vector by a number and returns the result.
func MultiplyNumber[T constraints.Integer | constraints.Float](a []T, b T) []T {
	result := make([]T, len(a))
	for i := range a {
		result[i] = a[i] * b
	}
	return result
}

// MultiplyNumberInPlace multiplies each element of a vector by a number and stores the result in the vector.
func MultiplyNumberInPlace[T constraints.Integer | constraints.Float](a []T, b T) {
	for i := range a {
		a[i] *= b
	}
}

// DivideNumber divides each element of a vector by a number and returns the result.
func DivideNumber[T constraints.Integer | constraints.Float](a []T, b T) []T {
	assertNotZero(b)
	result := make([]T, len(a))
	for i := range a {
		result[i] = a[i] / b
	}
	return result
}

// DivideNumberInPlace divides each element of a vector by a number and stores the result in the vector.
func DivideNumberInPlace[T constraints.Integer | constraints.Float](a []T, b T) {
	assertNotZero(b)
	for i := range a {
		a[i] /= b
	}
}

// Dot computes the dot product of two vectors.
func Dot[T constraints.Integer | constraints.Float](a, b []T) T {
	assertEqualLength(a, b)
	var result T = 0
	for i := 0; i < len(a); i++ {
		result += a[i] * b[i]
	}
	return result
}

// Magnitude computes the magnitude of a vector.
func Magnitude[T constraints.Integer | constraints.Float](a []T) T {
	return T(math.Sqrt(float64(Dot(a, a))))
}

// Normalize normalizes a vector to have a magnitude of 1.
func Normalize[T constraints.Integer | constraints.Float](a []T) []T {
	mag := Magnitude(a)
	return DivideNumber(a, mag)
}

// Sum computes the sum of all elements in a vector.
func Sum[T constraints.Integer | constraints.Float](a []T) T {
	var result T = 0
	for i := range a {
		result += a[i]
	}
	return result
}

// Product computes the product of all elements in a vector.
func Product[T constraints.Integer | constraints.Float](a []T) T {
	var result T = 1
	for i := 0; i < len(a); i++ {
		result *= a[i]
	}
	return result
}

// func Mean computes the mean of a vector.
func Mean[T constraints.Integer | constraints.Float](a []T) T {
	assertNonEmpty(a)
	return T(float64(Sum(a)) / float64(len(a)))
}

// Min computes the minimum element in a vector.
func Min[T constraints.Integer | constraints.Float](a []T) T {
	assertNonEmpty(a)
	result := a[0]
	for i := 0; i < len(a); i++ {
		if a[i] < result {
			result = a[i]
		}
	}
	return result
}

// Max computes the maximum element in a vector.
func Max[T constraints.Integer | constraints.Float](a []T) T {
	assertNonEmpty(a)
	result := a[0]
	for i := 0; i < len(a); i++ {
		if a[i] > result {
			result = a[i]
		}
	}
	return result
}

// Abs computes the absolute value of each element in a vector.
func Abs[T constraints.Integer | constraints.Float](a []T) []T {
	result := make([]T, len(a))
	for i := range a {
		result[i] = a[i]
		if a[i] < 0 {
			result[i] = -a[i]
		}
	}
	return result
}

// AbsInPlace computes the absolute value of each element in a vector and stores the result in the vector.
func AbsInPlace[T constraints.Integer | constraints.Float](a []T) {
	for i := range a {
		if a[i] < 0 {
			a[i] = -a[i]
		}
	}
}

// EuclidianDistance computes the Euclidian distance between two vectors.
func EuclidianDistance[T constraints.Integer | constraints.Float](a, b []T) float64 {
	assertEqualLength(a, b)
	var result float64 = 0
	for i := 0; i < len(a); i++ {
		result += math.Pow(float64(a[i]-b[i]), 2)
	}
	return math.Sqrt(result)
}

func assertEqualLength[T constraints.Integer | constraints.Float](a, b []T) {
	if len(a) != len(b) {
		panic("vectors must be the same length")
	}
}

func assertNonEmpty[T constraints.Integer | constraints.Float](a []T) {
	if len(a) == 0 {
		panic("vector must be non-empty")
	}
}

func assertNotZero[T constraints.Integer | constraints.Float](a T) {
	if a == 0 {
		panic("value must be non-zero")
	}
}
